### PR TITLE
[ws-manager] Introduce workspace classes

### PR DESCRIPTION
**Pull request by csweichel**
## Description
This PR adds support for workspace classes. Workspace classes add per-workspace support for different resource requests and limits, as well as different templates.

## Related Issue(s)

works on #8261 

## How to test
See unit tests

## Release Notes

```release-note
[ws-manager] Add support for workspace classes
```

### CODE REVIEW

File: components/ws-manager-api/go/config/config.go [hunk 1]
```diff
@@ -113,11 +109,13 @@ type Configuration struct {
...
+	// WorkspaceClasses provide different resource classes for workspaces
+	WorkspaceClasses map[string]*WorkspaceClass `json:"workspaceClass"`
 }
 
-// AllContainerConfiguration contains the configuration for all container in a workspace pod
-type AllContainerConfiguration struct {
-	Workspace ContainerConfiguration `json:"workspace"`
+type WorkspaceClass struct {
+	Container ContainerConfiguration            `json:"container"`
+	Templates WorkspacePodTemplateConfiguration `json:"templates"`
```
Comment (princerachit) [inline components/ws-manager-api/go/config/config.go:118]: maybe singular `Template`?

File: components/ws-manager-api/go/config/config.go [hunk 2]
```diff
@@ -206,23 +200,33 @@ func (c *Configuration) Validate() error {
...
+			validation.Field(&class.Templates.PrebuildPath, validPodTemplate),
+			validation.Field(&class.Templates.ProbePath, validPodTemplate),
+			validation.Field(&class.Templates.RegularPath, validPodTemplate),
+		)
+		if err != nil {
+			return xerrors.Errorf("workspace class %s: %w", name, err)
+		}
+	}
+
 	return err
```
Comment (princerachit) [inline components/ws-manager-api/go/config/config.go:230]: Is this redundant return? Should we return nil here?

File: components/ws-manager/pkg/manager/create.go [hunk 3]
```diff
@@ -42,20 +42,25 @@ var (
 // createWorkspacePod creates the actual workspace pod based on the definite workspace pod and appropriate
 // templates. The result of this function is not expected to be modified prior to being passed to Kubernetes.
 func (m *Manager) createWorkspacePod(startContext *startWorkspaceContext) (*corev1.Pod, error) {
-	podTemplate, err := config.GetWorkspacePodTemplate(m.Config.WorkspacePodTemplate.DefaultPath)
+	var templates config.WorkspacePodTemplateConfiguration
+	if startContext.Class != nil {
+		templates = startContext.Class.Templates
+	}
+
+	podTemplate, err := config.GetWorkspacePodTemplate(templates.DefaultPath)
```
Comment (princerachit) [inline components/ws-manager/pkg/manager/create.go:50]: Is it possible that `templates` is not initialized?

**Review by princerachit — COMMENTED** [review]
nit picks

File: components/ws-manager-api/go/config/config.go (hunk 1 above)
Comment (csweichel) [reply components/ws-manager-api/go/config/config.go:118]: The `WorkspacePodTemplateConfiguration` contains multiple templates, hence the plural

**Review by aledbf — APPROVED** [review]

**Review by Furisto — APPROVED** [review]

**Review by sagor999 — APPROVED** [review]

File: install/installer/pkg/config/v1/experimental/experimental.go [hunk 4]
```diff
@@ -58,6 +61,22 @@ type WorkspaceConfig struct {
 			PasswordSecret string   `json:"passwordSecret"`
 		} `json:"redisCache"`
 	} `json:"registryFacade"`
+
+	WorkspaceClasses map[string]WorkspaceClass `json:"classes,omitempty"`
```
Comment (Pothulapati) [inline install/installer/pkg/config/v1/experimental/experimental.go:65]: Shouldn't the `json` field be `workspaceClasses`? as `classes` seems a bit ambiguous? 🤔

**Review by Pothulapati — APPROVED** [review]
LGTM, except for the naming nit!

**Pull request merged**